zrml/prediction-markets/src/lib.rs

```diff
@@ -139,27 +151,47 @@ mod pallet {
             let market_status = market.status;
             let outcome_assets = Self::outcome_assets(market_id, &market);
             let outcome_assets_amount = outcome_assets.len();
-            Self::clear_auto_resolve(&market_id)?;
-            T::MarketCommons::remove_market(&market_id)?;
-            Self::deposit_event(Event::MarketDestroyed(market_id));
+            let market_account = Self::market_account(market_id);
 
-            let mut outcome_assets_iter = outcome_assets.into_iter();
+            // Slash outstanding bonds; see
+            // https://github.com/zeitgeistpm/runtime-audit-1/issues/34#issuecomment-1120187097 for
+            // details.
+            let slash_market_creator = |amount| {
+                CurrencyOf::<T>::slash_reserved_named(&RESERVE_ID, &market.creator, amount);
+            };
+            if market_status == MarketStatus::Proposed {
+                slash_market_creator(T::AdvisoryBond::get());
+            }
+            if market_status != MarketStatus::Resolved
+                && market_status != MarketStatus::InsufficientSubsidy
+            {
+                if market.creation == MarketCreation::Permissionless {
+                    slash_market_creator(T::ValidityBond::get());
+                }
+                slash_market_creator(T::OracleBond::get());
+            }
 
-            // Delete of this market's outcome assets.
-            let mut manage_outcome_asset = |asset: Asset<_>| -> usize {
+            // Delete market's outcome assets, clear market and delete pool if necessary.
+            let mut destroy_asset = |asset: Asset<_>| -> usize {
                 let (total_accounts, accounts) = T::Shares::accounts_by_currency_id(asset);
                 share_accounts = share_accounts.saturating_add(accounts.len());
                 T::Shares::destroy_all(asset, accounts.iter().cloned());
                 total_accounts
             };
-
-            if let Some(first_asset) = outcome_assets_iter.next() {
-                total_accounts = manage_outcome_asset(first_asset);
+            for asset in outcome_assets.into_iter() {
+                total_accounts = destroy_asset(asset);
             }
-            for asset in outcome_assets_iter {
-                let _ = manage_outcome_asset(asset);
+            CurrencyOf::<T>::slash(&market_account, CurrencyOf::<T>::free_balance(&market_account));
+            if let Ok(pool_id) = T::MarketCommons::market_pool(&market_id) {
+                T::Swaps::destroy_pool(pool_id)?;
```

Comment:
The Rikiddo markets need some additional treatment here:
1. If the market is in `CollectingSubsidy` state, the market has to be removed from the `MarketsCollectingSubsidy` vector, otherwise the pallet will try to process it at a later time
2. In Rikiddo pools the liquidity providers reserve their liquidity in ZTG before the pool becomes active. Once it becomes active, ZTG is converted to `PoolShares`. When a market that is in state `CollectingSubsidy` should be destroyed, those liquidity reserves from the so called "subsidy providers" must be freed (either by unreserving or in that case slashing them). The swaps pallet provides a function called `destroy_pool_in_subsidy_phase`. It should be called instead of `destroy_pool` for markets in `CollectingSubsidy` state. Feel free to add an additional parameter to that function that indicates whether unreserving or slashing should happen.

Reply:
(Same as below, I'd really like to skip Rikiddo for now)